phoenix-core/src/main/java/org/apache/phoenix/util/IndexUtil.java

```diff
@@ -849,4 +857,59 @@ public static long getIndexPendingDisableCount(PhoenixConnection conn, String fa
             throw new IOException(e);
         }
     }
+
+    private static boolean containsOneOrMoreColumn(Scan scan) {
+        Map<byte[], NavigableSet<byte[]>> familyMap = scan.getFamilyMap();
+        if (familyMap == null || familyMap.isEmpty()) {
+            return false;
+        }
+        for (Map.Entry<byte[], NavigableSet<byte[]>> entry : familyMap.entrySet()) {
+            NavigableSet<byte[]> family = entry.getValue();
+            if (family != null && !family.isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static void updateScan(Scan scan, byte[] emptyCF, byte[] emptyCQ) {
+        if (containsOneOrMoreColumn(scan)) {
+            scan.addColumn(emptyCF, emptyCQ);
+            return;
+        }
+        return;
+    }
+
+    public static void setScanAttributes(Scan scan, PTable table, PhoenixConnection phoenixConnection) throws SQLException {
+        if (table.isTransactional() || table.isImmutableRows() || table.getType() != PTableType.INDEX) {
+            return;
+        }
+        PTable indexTable = table;
+        if (indexTable.getIndexType() != PTable.IndexType.GLOBAL) {
+            return;
+        }
+        String schemaName = indexTable.getParentSchemaName().getString();
+        String tableName = indexTable.getParentTableName().getString();
+        PTable dataTable = PhoenixRuntime.getTable(phoenixConnection, SchemaUtil.getTableName(schemaName, tableName));
```

Comment:
PhoenixRuntime.getTable could return a cached copy of the PTable rather than re-creating from syscat. Is that OK? What if the index has been created / modified since the last cache load?

Reply:
I am not sure about the answer to your question. I agree that non-cached version would be safer but why do we have the cached version then and when to use cached versus non-cache? I did not want to impact the performance here. @twdsilva, do we have any guidelines on this?

Reply:
The cache would be on the client side to avoid hitting syscat on each read. I don't think we should not use the cache. Do we really need to get dataTable? I see it's usage for checking index really being on it (lines 919-921) and for PHYSICAL_DATA_TABLE_NAME. Can we simply not do the check on lines 919-921 and use tableName (from line 917)? Alternative solution - wouldn't the GlobalIndexChecker already have the index-table properties and know the data-table info? That way we can avoid the clients sending this needless info to server side; and be more fool-proof as well. [Imagine a rogue client sending a bad physical data-table name to corrupt the index]

Reply:
HBase coprocessors are stateless. They are just hooks. They do not have any info on the Phoenix tables. Also GlobalIndexChecker is added only to the index tables.